### ISSUE TITLE
android sqlcipher 4.4.0 query functions has a serious bug

### ISSUE THREAD

**Issue by jiangmingliu** [sqlcipher/android-database-sqlcipher#526]
# Steps to Reproduce
1.  an mobile app(android apk) use android-sqlciper 4.4.0 version ;
2.  use the sql query Statement with inner join， for example：
## NOTES: we have two talbes named sample_session, sample_point, the table structure is start_time, client_id, type_id, value; ##

SELECT sample_session._id, sample_session.start_time, sample_session.end_time, sample_session.type_id, sample_session.metadata
	, sample_session.client_id, sample_session.timezone, sample_session.modified_time, SUM(CASE sample_point.type_id
		WHEN 2 THEN sample_point.value
		ELSE 0
	END) AS step
	, SUM(CASE sample_point.type_id
		WHEN 4 THEN sample_point.value
		ELSE 0
	END) AS calorie, SUM(CASE sample_point.type_id
		WHEN 3 THEN sample_point.value
		ELSE 0
	END) AS distance
	, SUM(CASE sample_point.type_id
		WHEN 5 THEN sample_point.value
		ELSE 0
	END) AS altitude_offset
FROM sample_session
	INNER JOIN sample_point
	ON sample_session.start_time = sample_point.start_time
		AND sample_session.client_id = sample_point.client_id
WHERE sample_session.sync_status = 0
	AND sample_session.client_id = 5
	AND sample_point.client_id = 5
	AND sample_point.type_id IN (2, 4, 3, 5)
	AND sample_session.start_time = 1605502380000
GROUP BY sample_session.start_time
ORDER BY sample_session.start_time DESC
LIMIT 0, 1000

### Actual Behavior
1. the result contains the client_id = 6 and client_id = 5, 
2. when we use the sqlcipher version 4.2.0, the behavior becomes ok, the result ony include client_id = 5;
SQLCipher for Android version: 4.4.0

SQLCipher version (can be identified by executing `PRAGMA cipher_version;`):

*Note:* If you are not posting a specific issue for the SQLCipher library, please consider posting your question to the SQLCipher [discuss site](https://discuss.zetetic.net/c/sqlcipher). Thanks!

**Comment by jiangmingliu** [comment]
@developernotes

**Comment by developernotes** [comment]
Hello @jiangmingliu 

Would you move this into a reproduction test case that can be publicly reviewed for further analysis? You can do so within the SQLCipher for Android [test suite](https://github.com/sqlcipher/sqlcipher-android-tests). Thanks!

**Comment by jiangmingliu** [comment]
thanks your reply @developernotes 
actually which i wrote the sql statement in this issue， is from our app real sql statement，i can also add the particular test, but from my point. my detail info has been pointed out, that is the error sql statement.
we have two talbes named sample_session, sample_point, the table structure is start_time, client_id, type_id, value;
SELECT sample_session._id, sample_session.start_time, sample_session.end_time, sample_session.type_id, sample_session.metadata
, sample_session.client_id, sample_session.timezone, sample_session.modified_time, SUM(CASE sample_point.type_id
WHEN 2 THEN sample_point.value
ELSE 0
END) AS step
, SUM(CASE sample_point.type_id
WHEN 4 THEN sample_point.value
ELSE 0
END) AS calorie, SUM(CASE sample_point.type_id
WHEN 3 THEN sample_point.value
ELSE 0
END) AS distance
, SUM(CASE sample_point.type_id
WHEN 5 THEN sample_point.value
ELSE 0
END) AS altitude_offset
FROM sample_session
INNER JOIN sample_point
ON sample_session.start_time = sample_point.start_time
AND sample_session.client_id = sample_point.client_id
WHERE sample_session.sync_status = 0
AND sample_session.client_id = 5
AND sample_point.client_id = 5
AND sample_point.type_id IN (2, 4, 3, 5)
AND sample_session.start_time = 1605502380000
GROUP BY sample_session.start_time
ORDER BY sample_session.start_time DESC
LIMIT 0, 1000

**Comment by developernotes** [comment]
Hi @jiangmingliu 

It would be helpful for you to [create a new test](https://github.com/sqlcipher/sqlcipher-android-tests#creating-a-new-test) using the SQLCipher for Android test suite providing a sample database and your query so we can verify the behavior on both 4.4.0 and 4.2.0. This will also allow us to investigate the details further. Thanks!

**Comment by jiangmingliu** [comment]
@developernotes here is my test add in ur project，i get the result as i said before. i use the version 4.4.1 will be ok, but the 4.4.0 will failed. i hope you can give me more details to explain this bug.
```java
public class AwesomeTest extends SQLCipherTest {
    @Override
    public boolean execute(SQLiteDatabase database) {
        database.execSQL("create table sample_session(_id , start_time TEXT , client_id TEXT, type_id TEXT, value TEXT);");
        database.execSQL("create table sample_point(_id ,start_time TEXT , client_id TEXT, type_id TEXT, value TEXT);");
        database.execSQL("insert into sample_session(_id , start_time ,client_id, type_id, value) values(1, 1000,5 ,2, 0);");
        database.execSQL("insert into sample_point(_id , start_time ,client_id, type_id, value) values(1, 1000,5 ,2, 46);");
        database.execSQL("insert into sample_point(_id , start_time ,client_id, type_id, value) values(2, 1000,6 ,2, 46);");
        String query = "select sample_session._id,sample_session.start_time,sample_session.type_id, sample_session.client_id," +
                "sum ( case sample_point.type_id when 2 then sample_point.value else 0 end) as step, sum ( case sample_point.type_id when 4 then sample_point.value else 0 end) as calorie, sum ( case sample_point.type_id when 3 then sample_point.value else 0 end) as distance, " +
                "sum ( case sample_point.type_id when 5 then sample_point.value else 0 end) as altitude_offset from sample_session INNER JOIN sample_point ON sample_session.start_time = sample_point.start_time and sample_session.client_id = sample_point.client_id " +
                "where sample_session.client_id =? and sample_point.client_id =?  and sample_point.type_id in ( ?,?,?,? ) " +
                "group by sample_session.start_time ORDER BY sample_session.start_time DESC  limit 0,1000";
        Cursor cursor = database.rawQuery(query, new String[]{"5","5","2","4","3","5"});
        if(cursor != null){
            cursor.moveToFirst();
            int index = cursor.getColumnIndex("step");
            long value = cursor.getInt(index);
            return value == 46;
        }
        return false;
    }

    @Override
    public String getName() {
        return "Awesome Test";
    }
}
```
```Gradle
dependencies {
  // For testing JAR-based distribution:
  // implementation files('libs/sqlcipher.jar')

  // For testing local AAR packages:
  //implementation (name: 'android-database-sqlcipher-4.4.1-release', ext: 'aar')

  // For testing on remote AAR references:
  implementation 'net.zetetic:android-database-sqlcipher:4.4.0@aar'

  implementation "androidx.sqlite:sqlite:2.0.1"

  // For Room tests:
  implementation "androidx.room:room-runtime:2.1.0"
  annotationProcessor "androidx.room:room-compiler:2.1.0"
}
```

**Comment by jiangmingliu** [comment]
@developernotes

**Comment by jiangmingliu** [comment]
@developernotes btw，what is ur timezone，this issue is kind of urgent.BTW, does this bug be fixed in 4.4.1?

**Comment by developernotes** [comment]
Hi @jiangmingliu 

Have you tried updating your library reference to 4.4.1? It appears to return the results you are looking for whereas 4.4.0 doesn't. Please try updating your library to the latest.

![4 4 1-test-results](https://user-images.githubusercontent.com/3225/99543625-fcdba700-2978-11eb-9f1a-baf19b22f8c5.png)
![4 4 0-test-results](https://user-images.githubusercontent.com/3225/99543632-ff3e0100-2978-11eb-8ec8-02e400ac123e.png)

**Comment by jiangmingliu** [comment]
@developernotes , sure, i have already metioned that the 4.4.1 version will works well. I  hope you can give me a detailed analysis based on ur source code, so that i evaluate the impact on the version of 4.4.0, and also you can double confirm the problem has been fixed in 4.4.1 version.

**Comment by jiangmingliu** [comment]
@developernotes without the root cause analysis,  i am not sure whether this lastest version has been completely fix the problem i met. i wish you nice guys spare some time to give an answer.

**Comment by jiangmingliu** [comment]
hi @developernotes , it would be helpful you guys can give me a reply, it is really important for me,  Thank you, please

**Comment by jiangmingliu** [comment]
hi @developernotes again, still looking forward to ur reply...how was ur weekend

**Comment by sjlombardo** [comment]
Hello @jiangmingliu - thanks for following up on this. We are working on determining the cause of the problem. We do not yet have an update. We will update the ticket as soon as we have more information.

**Comment by jiangmingliu** [comment]
@sjlombardo really appreciate that.

**Comment by jiangmingliu** [comment]
hi @developernotes @sjlombardo, do you have any infos to share about this issue?

**Comment by sjlombardo** [comment]
Hello @jiangmingliu - we have completed an analysis of this issue. This problem was present in upstream SQLite 3.31, the version of SQLite that SQLCipher 4.4.0 was based on. The underlying issue was a bug in SQLite related to the combination of where clause constraints. The bug was subsequently fixed SQLite 3.32 and included in the newer releases of SQLCipher. Our recommendation would be to use SQLCipher 4.4.2.

**Comment by jiangmingliu** [comment]
how do i confirm the version of sqlite from the aar file ? on the other hand, what is the sqlite version in the version of 4.4.1. because i found it works well in the version in 4.4.1.  And that is  vague **a bug in SQLite related to the combination of where clause constraints**, it would be helpful to give me more details.

**Comment by jiangmingliu** [comment]
@sjlombardo

**Comment by sjlombardo** [comment]
Hello @jiangmingliu - You should look at the SQLCipher [Changelog](https://github.com/sqlcipher/sqlcipher/blob/master/CHANGELOG.md) to determine which SQLite versions have been used in each SQLCipher release. Each time we merge in an SQLite update it is noted there. SQLCipher 4.4.2 is based on SQLite 3.33.0. You can also check the baseline SQLite version at runtime by executing `SELECT sqlite_version();`.  

In terms of separate verification, you can confirm the bug and it's resolution yourself by downloading and building standard SQLite for the versions referenced. You'll note that the issue was present in 3.31.x and fixed in 3.32. 

If you are interested in additional support, further details, or more thorough analysis you are welcome to contact us at support@zetetic.net about commercial support offerings.

**Comment by jiangmingliu** [comment]
i see ... @sjlombardo thank you very much.
